File: exp12.py
import sys

class TravellingSalesmanProblem:
    def __init__(self, graph):
        """
        Initialize the TSP problem with a given graph.
        
        Parameters:
            graph: A 2D list representing the adjacency matrix of the graph.
        """
        self.graph = graph
        self.n = len(graph)
        self.min_cost = sys.maxsize
        self.best_path = []

    def tsp_branch_and_bound(self):
        """
        Solves the TSP using the Branch and Bound approach.
        
        Returns:
            The minimum cost and the path of the optimal solution.
        """
        # Initialize variables
        visited = [False] * self.n
        visited[0] = True  # Start from the first city
        self._branch_and_bound(0, visited, 1, 0, [0])
        return self.min_cost, self.best_path

    def _branch_and_bound(self, current_city, visited, count, current_cost, current_path):
        """
        Recursive helper function for the Branch and Bound approach.
        
        Parameters:
            current_city: The city currently being visited.
            visited: A list of booleans indicating whether each city has been visited.
            count: Number of cities visited so far.
            current_cost: Cost of the current path.
            current_path: The current path being explored.
        """
        # Base case: If all cities are visited, return to the starting city
        if count == self.n and self.graph[current_city][0] != 0:
            total_cost = current_cost + self.graph[current_city][0]
            if total_cost < self.min_cost:
                self.min_cost = total_cost
                self.best_path = current_path + [0]
            return

        # Explore all possible cities from the current city
        for next_city in range(self.n):
            if not visited[next_city] and self.graph[current_city][next_city] != 0:
                # Calculate the cost of visiting the next city
                new_cost = current_cost + self.graph[current_city][next_city]

                # Prune the branch if the cost exceeds the current minimum
                if new_cost < self.min_cost:
                    visited[next_city] = True
                    self._branch_and_bound(next_city, visited, count + 1, new_cost, current_path + [next_city])
                    visited[next_city] = False  # Backtrack

# Example usage
if __name__ == "__main__":
    # Example adjacency matrix (distance between cities)
    graph = [
        [0, 10, 15, 20],
        [10, 0, 35, 25],
        [15, 35, 0, 30],
        [20, 25, 30, 0]
    ]

    tsp = TravellingSalesmanProblem(graph)
    min_cost, best_path = tsp.tsp_branch_and_bound()
    print("Minimum Cost:", min_cost)
    print("Optimal Path:", best_path)

File: exp12.c
#include <stdio.h>
#include <limits.h>
#include <stdbool.h>

#define MAX 20 // Maximum number of cities

int n; // Number of cities
int cost[MAX][MAX]; // Cost matrix
int visited[MAX]; // Visited cities
int minCost = INT_MAX; // Minimum cost found
int path[MAX]; // Path taken

// Function to find the minimum cost path using Branch and Bound
void tsp(int currentPos, int count, int costSoFar) {
    // If all cities are visited and we return to the starting city
    if (count == n && cost[currentPos][0]) {
        if (costSoFar + cost[currentPos][0] < minCost) {
            minCost = costSoFar + cost[currentPos][0];
            path[count] = 0; // Return to starting city
        }
        return;
    }

    // Explore all possible paths
    int i;
    for (i = 0; i < n; i++) {
        // If the city is not visited and there is a path
        if (!visited[i] && cost[currentPos][i]) {
            // Mark the city as visited
            visited[i] = true;
            path[count] = i;

            // Recur to the next city
            tsp(i, count + 1, costSoFar + cost[currentPos][i]);

            // Backtrack
            visited[i] = false;
        }
    }
}

int main() {
    printf("Enter the number of cities: ");
    scanf("%d", &n);

    printf("Enter the cost matrix:\n");
    int i, j;
    for (i = 0; i < n; i++) {
        for (j = 0; j < n; j++) {
            scanf("%d", &cost[i][j]);
        }
    }

    // Initialize visited array
    for (i = 0; i < n; i++) {
        visited[i] = false;
    }

    // Start from the first city
    visited[0] = true;
    path[0] = 0;

    // Call the TSP function
    tsp(0, 1, 0);

    // Print the minimum cost and path
    printf("Minimum cost: %d\n", minCost);
    printf("Path: ");
    for (i = 0; i <= n; i++) {
        printf("%d ", path[i]);
    }
    printf("\n");

    return 0;
}

